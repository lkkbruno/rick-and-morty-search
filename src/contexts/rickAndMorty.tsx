import React, { createContext, useState, useEffect, PropsWithChildren } from 'react';
import axios from 'axios';

export interface ICharacter {
  id: number;
  name: string;
  status: string;
  species: string;
  type: string;
  gender: string;
  origin: {
    name: string;
    url: string;
  };
  location: {
    name: string;
    url: string;
  };
  image: string;
  episode: string[];
  url: string;
  created: string;
}

interface RickAndMortyContextData {
  characters: ICharacter[];
  searchTerm: string;
  setSearchTerm: (term: string) => void;
}

export const RickAndMortyContext = createContext<RickAndMortyContextData>({
  characters: [],
  searchTerm: '',
  setSearchTerm: () => { },
});

export const RickAndMortyProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [characters, setCharacters] = useState<ICharacter[]>([]);
  const [searchTerm, setSearchTerm] = useState<string>('');

  const fetchCharacters = () => {
    axios
      .get(`https://rickandmortyapi.com/api/character`, {
        params: {
          name: searchTerm,
        }
      })
      .then((response) => {
        setCharacters(response.data.results);
      }).catch(() => {
        setCharacters([]);
      });
  }

  useEffect(() => {
    fetchCharacters();
  }, [searchTerm]);

  return (
    <RickAndMortyContext.Provider value={{ characters, searchTerm, setSearchTerm }}>
      {children}
    </RickAndMortyContext.Provider>
  );
};
