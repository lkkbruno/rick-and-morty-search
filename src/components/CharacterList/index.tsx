import React, { useContext, useState, useEffect } from 'react';
import styled from 'styled-components';
import { RickAndMortyContext, ICharacter } from '../../contexts/rickAndMorty';
import Character from '../Character';

const List = styled.ul`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  list-style: none;
  padding: 0;
  margin: 0;
`;

interface CharacterListProps {
  characters: ICharacter[],
}

const CharacterList: React.FC<CharacterListProps> = ({ characters }) => {
  const { searchTerm } = useContext(RickAndMortyContext);
  const [filteredCharacters, setFilteredCharacters] = useState<ICharacter[]>(characters);

  useEffect(() => {
    setFilteredCharacters(
      characters.filter((character) =>
        character.name.toLowerCase().includes(searchTerm.toLowerCase())
      )
    );
  }, [characters, searchTerm]);

  return (
    <List>
      {filteredCharacters.map((character) => (
        <Character key={character.id} character={character} />
      ))}
    </List>
  );
};

export default CharacterList;
