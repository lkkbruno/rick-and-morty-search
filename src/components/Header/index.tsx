import React, { ChangeEvent, useContext, useState } from 'react';
import styled from 'styled-components';
import { RickAndMortyContext } from '../../contexts/rickAndMorty';
// @ts-ignore
import logo from '../../assets/rickAndMortyLogo.png';

const Container = styled.header`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  padding: 16px;
`;

const Logo = styled.img`
  width: 40%;
  margin-bottom: 20px;
`;

const SearchInput = styled.input`
  height: 40px;
  width: 88vw;
  border-radius: 25px;
  font-size: 16px;
  font-weight: 400;
  padding: 4px 50px;
`;

interface HeaderProps {
  searchTerm: string;
  setSearchTerm: (searchTerm: string) => void;
}

const Header: React.FC<HeaderProps> = ({ searchTerm, setSearchTerm }) => {
  const [inputValue, setInputValue] = useState<string>(searchTerm);

  const handleSearchTermChange = (event: ChangeEvent<HTMLInputElement>) => {
    const newSearchTerm = event.target.value;
    setInputValue(newSearchTerm);
    setSearchTerm(newSearchTerm);
  };

  return (
    <Container>
      <Logo src={logo} alt="Rick and Morty" />
      <SearchInput
        type="text"
        placeholder="Search characters by name"
        value={inputValue}
        onChange={handleSearchTermChange}
        data-testid="search-input"
      />
    </Container>
  );
};

export default Header;
