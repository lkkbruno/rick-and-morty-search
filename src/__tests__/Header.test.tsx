import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { RickAndMortyProvider } from '../contexts/rickAndMorty';
import Header from '../components/Header';
import CharacterList from '../components/CharacterList';

describe('Header', () => {
  it('should match snapshot', async () => {
    const { container } = render(<Header searchTerm='' setSearchTerm={() => { }} />)
    expect(container).toMatchSnapshot();
  })
  it('renders the logo', () => {
    const { getByAltText } = render(
      <Header searchTerm='' setSearchTerm={() => { }} />
    );

    expect(getByAltText(/Rick and Morty/i)).toBeInTheDocument();
  });

  it('renders the search input', () => {
    const { getByTestId } = render(
      <Header searchTerm='' setSearchTerm={() => { }} />
    );

    expect(getByTestId('search-input')).toBeInTheDocument();
  });

  it('updates the search term on input change ', async () => {
    const { getByPlaceholderText, findByText } = render(
      <Header searchTerm='' setSearchTerm={() => { }} />
    );

    const input = getByPlaceholderText(/search characters by name/i);

    expect(input).toHaveValue('');
    fireEvent.change(input, { target: { value: 'Rick' } });
    expect(input).toHaveValue('Rick');
  });
});
