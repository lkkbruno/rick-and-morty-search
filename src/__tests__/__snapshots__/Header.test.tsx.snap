// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match snapshot 1`] = `
<div>
  <header
    class="sc-aXZVg fymyJh"
  >
    <img
      alt="Rick and Morty"
      class="sc-gEvEer jLCWPF"
      src=""
    />
    <input
      class="sc-eqUAAy lbFcbm"
      data-testid="search-input"
      placeholder="Search characters by name"
      type="text"
      value=""
    />
  </header>
</div>
`;
